lib/forklift-server/src/middleware/consumer_lag_gauge.rs

```diff
@@ -0,0 +1,181 @@
+            emit_fn: self.emit_fn.clone(),
+            time_threshold: self.time_threshold,
+            message_threshold: self.message_threshold,
+            last_report: Arc::new(Mutex::new(Instant::now())),
+            messages_since_report: Arc::new(AtomicU64::new(0)),
+        }
+    }
+}
+
+/// Service that emits consumer lag gauge periodically
+#[derive(Clone)]
+pub(crate) struct ConsumerLagGauge<S, F> {
+    inner: S,
+    emit_fn: Arc<F>,
+    time_threshold: Duration,
+    message_threshold: u64,
+    last_report: Arc<Mutex<Instant>>,
+    messages_since_report: Arc<AtomicU64>,
+}
+
+impl<S, F> Service<Message<jetstream::Message>> for ConsumerLagGauge<S, F>
+where
+    S: Service<Message<jetstream::Message>> + Clone + Send + 'static,
+    S::Future: Send + 'static,
+    F: Fn(u64) + Send + Sync + 'static,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>> + Send>>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, req: Message<jetstream::Message>) -> Self::Future {
+        // Extract lag from message metadata before passing to inner service
+        let lag = req.info().ok().map(|info| info.pending).unwrap_or(0);
+
+        // Track messages and check if we should emit
+        let messages_processed = self.messages_since_report.fetch_add(1, Ordering::Relaxed) + 1; // +1 because fetch_add returns previous value
+
+        let last_report = self.last_report.clone();
+        let messages_since_report = self.messages_since_report.clone();
+        let emit_fn = self.emit_fn.clone();
+        let time_threshold = self.time_threshold;
+        let message_threshold = self.message_threshold;
+
+        // Check thresholds and emit if needed
+        let emit_future = async move {
+            let mut should_emit = false;
+            let mut last = last_report.lock().await;
+
+            // Check time threshold
+            if last.elapsed() >= time_threshold {
+                should_emit = true;
+            }
+
+            // Check message count threshold
+            if messages_processed >= message_threshold {
+                should_emit = true;
+            }
+
+            if should_emit {
+                // Call the emission function with the lag value
+                (emit_fn)(lag);
+
+                // Reset tracking
+                *last = Instant::now();
+                messages_since_report.store(0, Ordering::Relaxed);
+            }
+        };
+
+        // Clone the service for the async block
+        let clone = self.inner.clone();
+        let mut inner = std::mem::replace(&mut self.inner, clone);
+
+        Box::pin(async move {
+            // Emit gauge if threshold met
+            emit_future.await;
+
+            // Pass request to inner service
+            inner.call(req).await
+        })
+    }
```

Comment:
Let me see if I understand this correctly: the lag guage middleware will not block message processing because the emit future and the inner request pass are both awaited _**but**_ that happens inside a pinned, moved async function? What if the emit future never returns?

Reply:
Oh yeah, that was a little different of a setup.

The middleware is generally optionally running some code before the "inner", then optionally running code after. The `tower::Service` trait we're using everywhere in naxum (also in axum) has a non-async method of `fn call(...) -> Self::Future` so if the before code is sync code, you can call that directly, then return an `impl Future` for the "inner" as the "inner" is async code.

In this case, the code contained in the `emit_future` block had an await point on a lock, so it was easier to put that code into the returning `impl Future`. The lock in this case is a Tokio `Mutex` so as I recall a lock requests are queued in the order of request and this lock doesn't poison (unlike the `Mutex` from stdlib).